src/acl/Acl.cc

```diff
@@ -288,15 +288,23 @@ ACL::isProxyAuth() const
 void
 ACL::parseFlags()
 {
-    // ACL kids that carry ACLData which supports parameter flags override this
-    Acl::ParseFlags(options(), Acl::NoFlags());
+    Acl::Options allOptions = options();
+    for (const auto lineOption: lineOptions()) {
+        lineOption->unconfigure(); // forget any previous "acl ..." line effects
```

Comment:
We should not. The method does not reset or clear the object! Some of the object settings are (and should be) preserved. Roughly speaking, the method only clears the data collected on the previous ACL line.

Reply:
Your responses confirm what I thought. This is a *reset* operation. Reset means revert the option(s) to their default values - it does *not* means erasing configuration or anything beyond value of the option.

Reply:
> Reset means revert the option(s) to their default values - it does not means erasing configuration or anything beyond value of the option.

The word "reset" means a lot of things to a lot of people in various contexts. Here, we are talking about a C++ class method. In STL and similar C++ contexts that many Squid developers should be familiar with, a reset() method means, roughly speaking, returning the object to the initially constructed (often default-constructed) object state. The unconfigure() method does not and cannot return the option object to that initial state. The Option object has a lot more state than "value", and this method must preserve a big chunk of that state. Calling it reset() or clear() would be very misleading to most readers familiar with STL.

Reply:
> Eduard: we only 'clear' the option value, leaving its name(s) intact.

To avoid misunderstanding, the above is correct but incomplete: The name(s) are not the only parts of the Option object (in aggregate) that unconfugure() does not reset. This API/method does not have a reset() or clear() semantics, and one of its current implementations would violate that reset() or clear() semantics if renamed.

Reply:
I think you mean the child objects inheriting from Option have a lot of state. When this method is implemented it is a wrapper around reset() on the location the value is stored.

This whole call sequence of ParseFlags() having to reset indirect storage locations on every call smells badly. But I am not going to insist on fixing that pre-existing weirdness here.

Reply:
Thank you for resolving this change request. Please feel free to ignore the clarifications below.

----

> I think you mean the child objects inheriting from Option have a lot of state. 

When naming an Option method, we should be thinking about the Option class/API as a whole, including the needs of any valid implementations of that abstract class, especially when the method is pure virtual. That is what I was naturally doing. The abstract Options class itself does not even store a value -- there is no need for a reset() or unconfigure() method at all as far as that specific class is concerned. Those concepts surface only in the context of that API users and implementors.

> When this method is implemented it is a wrapper around reset() on the location the value is stored.

Yes, TypedOption::unconfigure() calls `recipient_->reset()`, but that is actually _unimportant_ in the context of this discussion. The important part is what that TypedOption::unconfigure() method is _not_ doing! It is not resetting or clearing the `recipient_` data member. A proper reset() or clear() method implementation would have made `recipient_` nil because that part of the TypedOption state is temporary -- it changes from one `acl` name to another while the TypedOption object survives that transition (to keep option names and other construction-time state).


> This whole call sequence of ParseFlags() having to reset indirect storage locations on every call smells badly. But I am not going to insist on fixing that pre-existing weirdness here.

Yes, this performance optimization is ugly. I hope that someday, somebody will prove that it is not necessary -- we can probably just create and configure new Option objects for every `acl...` line instead.

And we can do more to make this code comprehensible -- see the second (out of three) parts of https://github.com/measurement-factory/squid/commit/fd11bbd0513f35158f2512677bfda9897db5568f commit message for the discussion of the primary complexity source and a possible long-term solution.

All out of scope here, of course.